src/Application/Actions/Donations/Create.php

```diff
@@ -279,4 +281,34 @@ private function removeSpecialChars(string $descriptor): string
     {
         return preg_replace('/[^A-Za-z0-9 ]/', '', $descriptor);
     }
+
+    /**
+     * It seems like just the *first* persist of a given donation needs to be retry-safe, since there is a small
+     * but non-zero minority of Create attempts at the start of a big campaign which get a closed Entity Manager
+     * and then don't know about the connected #campaign on persist and crash when RetrySafeEntityManager tries again.
+     */
+    private function persistDonationWithRetry(Donation $donation): void
+    {
+        \assert($this->entityManager instanceof RetrySafeEntityManager);
+        $retryCount = 0;
+        while ($retryCount < self::MAX_CREATE_RETRY_COUNT) {
+            try {
+                $this->entityManager->persistWithoutRetries($donation);
+                $this->entityManager->flush();
+                return;
+            } catch (ORMException $exception) {
+                $retryCount++;
+                $this->logger->info(
+                    sprintf(
+                        'Donation Create persist error: %s. Retrying %d of %d.',
+                        $exception->getMessage(),
+                        $retryCount,
+                        self::MAX_CREATE_RETRY_COUNT,
+                    )
+                );
+
+                usleep(random_int(100_000, 1_100_000)); // Wait between 0.1 and 1.1 seconds before retrying
```

Comment:
So as I understand it the only differences between this retry and the one built into the RetrySafeEntityManager are:
- there's a sleep
- We *do not* replace the underlying entity manager. This means if it's still usable we still have any required related new entities in Unit of Work - but the underlying EM is closed the retry will not work.

Is that right? Maybe worth adding a comment to that effect here?

Reply:
Yes, I think so. I'm still not totally confident about the ability to recover from a lost EM but have tried to combine this explanation with the existing method doc block in https://github.com/thebiggive/matchbot/pull/668/commits/ad6312283e74eca2f85f4bbe33ce0f55d89ed634